### PR TITLE
Instrumentation of node-rdfafka/kafka-avro

### PR DESCRIPTION
**Pull request by willianpc**
### Node-RdKafka Limitations

These limitations impact how the tracer performs and must be explicit to customers in the documentation:

 * The **Producer as a stream** can only have span correlation if the Writable option **objectMode is set to true**, otherwise, there is no way to append the Instana headers to it.
 * The Producer, as stream or standard API **cannot** propagate span correlation when headerFormat is set to `binary`. More info here: https://github.com/Blizzard/node-rdkafka/pull/935.
 * If the option `dr_cb` is not set to true, we assume that the message was sent, but without any guarantee.

### Relevant Notes

* Kafka-Avro currently works up until Node.js v14.17 due to a dependency to an old version of Node-RdKafka.
* Because of the above, we install kafka-avro as an optional dependency. However, even in Node.js v16 and above, npm will attempt to install it, but the installation won't fail if an error occurs. This means two things:
   * The CI time will increase also for Node v16 > as there will be actually an installation attempt.
   * There may be some garbage in node_modules after the installation attempt. Some packages installed during the installation that fails seem to not being cleaned up.
 * In order to test kafka-avro, we had to add Schema Registry to docker-compose.yml and config.yml, also increasing the time of the CI build.
 * Kafka configuration was changed in order to support and communicate properly with the Schema Registry server
 
 Tasks

 - [x] increase code coverage for error cases @kirrg001 
 - [x] optimise readability in instrumention @kirrg001 
 - [x] add avro installation to the test for v14 @kirrg001 
 - [x] fix level !== 1 @kirrg001 
 - [x] objectMode @kirrg001 
 - [x] fix https://github.com/instana/nodejs/pull/486#discussion_r818908700 @kirrg001 
 - [x] fix https://github.com/instana/nodejs/pull/486#discussion_r818827112

### CODE REVIEW

File: packages/collector/test/tracing/messaging/kafka-avro/producer.js
Comment (kirrg001) [inline packages/collector/test/tracing/messaging/kafka-avro/producer.js:10]: Notes from call:

Because of the [binary bug in rdkafka](https://github.com/Blizzard/node-rdkafka/pull/935) the span correlation wouldn't work at all in binary mode (which is still the default mode in kafka instrumention)

Node.js tracer has already added support for kafka string headers although the kafka header migration has not even started yet (or is not finished yet).
Setting the `headerFormat` to 'string' here means, rdkafka/avro instrumention works if the consumer tracer and the producer tracer support and use string format.

Therefor the customer needs to set `headerFormat: string`.

@willianpc as discussed: `headerFormat` is not documented yet in our public docs. We might want to wait for @basti1302 opinion here

File: package-lock.json
Comment (kirrg001) [inline package-lock.json:107]: I am getting a diff for package-lock.json file on Node v16.

File: packages/core/src/tracing/instrumentation/messaging/rdkafka.js
Comment (kirrg001) [inline packages/core/src/tracing/instrumentation/messaging/rdkafka.js:42]: _Suggestion_: Use named functions

Original: [Function: KafkaConsumer]
Mocked: [Function (anonymous)]

File: packages/collector/test/tracing/messaging/node-rdkafka/test.js
Comment (kirrg001) [inline packages/collector/test/tracing/messaging/node-rdkafka/test.js:68]: We are currently testing for _"tracing enabled, no suppression"_ only with `RDKAFKA_PRODUCER_DELIVERY_CB: true`. I was wondering if we also want to test with `false`. My initial expectation here was that we have to iterate over the `producerEnableDeliveryCbOptions`.

File: packages/collector/test/tracing/messaging/node-rdkafka/test.js
Comment (kirrg001) [inline packages/collector/test/tracing/messaging/node-rdkafka/test.js:88]: Is it possible that something is missing in the tests regarding error cases?

See line https://github.com/instana/nodejs/pull/486/files#diff-1b77510f78eb55910153dd198bee01c7d0e63aa32d30da88adae14562db0d5f8R143

`withError` is never `receiver` as far as I can see.

Furthermore, there are three cases where we set `span.ec` in the instrumentation, but it looks like only [one case](https://github.com/instana/nodejs/pull/486/files#diff-24bbb2057190b9840c5c35b03ff3849adf0501722464d3ba476378bf0dfd8fd4R162) is covered.

File: packages/core/src/tracing/instrumentation/messaging/rdkafka.js
Comment (kirrg001) [inline packages/core/src/tracing/instrumentation/messaging/rdkafka.js:215]: I was wondering why we do not call the original emit function for both cases:

- isActive is false (https://github.com/instana/nodejs/pull/486/files#diff-24bbb2057190b9840c5c35b03ff3849adf0501722464d3ba476378bf0dfd8fd4R211)
- parentSpan is not null (https://github.com/instana/nodejs/pull/486/files#diff-24bbb2057190b9840c5c35b03ff3849adf0501722464d3ba476378bf0dfd8fd4R215)

**Comment by kirrg001** [comment]
Thanks for this great work 💯 🥳

File: packages/core/src/tracing/instrumentation/messaging/rdkafka.js
Comment (kirrg001) [inline packages/core/src/tracing/instrumentation/messaging/rdkafka.js:111]: _General question_: What is the use case for parentSpan being already an exit span?
So there is an incoming message, which should be produced to a topic, but the parent span is already an exit span.
When does that happen?

Thanks!

File: packages/collector/test/tracing/messaging/node-rdkafka/test.js
Comment (willianpc) [reply packages/collector/test/tracing/messaging/node-rdkafka/test.js:68]: Good catch. Will do it.

File: packages/collector/test/tracing/messaging/node-rdkafka/test.js
Comment (willianpc) [reply packages/collector/test/tracing/messaging/node-rdkafka/test.js:88]: > withError is never receiver as far as I can see

The reason why `receiver` is never set is because I could not find a way to force an error when receiving a message. Yet I kept the option in the test in case we find ways to reproduce errors in the future.

>  there are three cases where we set span.ec in the instrumentation

The uncovered cases include either an event emit from deliveryCb or a received message. We can discuss together, but as far as I could investigate, I could not properly force/capture errors in these cases

File: packages/core/src/tracing/instrumentation/messaging/rdkafka.js
Comment (willianpc) [reply packages/core/src/tracing/instrumentation/messaging/rdkafka.js:215]: Nice catch! I don't know why this did not cause errors before, as it should.
But they were missing for sure.
Thanks!

File: packages/core/src/tracing/instrumentation/messaging/rdkafka.js
Comment (willianpc) [reply packages/core/src/tracing/instrumentation/messaging/rdkafka.js:111]: Essentially, the idea is to check if there isn't already a Kafka exit span being processed, or any kind of exit span for that matter.
As a rule of thumb, there should never be an ENTRY -> ENTRY span correlation, or EXIT -> EXIT. The expectation is to always have the alternated ENTRY -> EXIT -> ENTRY and so on.

If this is not clear enough, please let me know :)

File: package.json
Comment (willianpc) [reply package.json:175]: That's a valid point. The slowness is significant.
Apart from that, as I mentioned in the PR description, since the installation of kafka-avro is optional, even where the installation is not possible, npm goes forward with the installation as far as it works, which causes an unnecessary amount of time spent for nothing.

File: packages/core/src/tracing/instrumentation/messaging/rdkafka.js
Comment (kirrg001) [reply packages/core/src/tracing/instrumentation/messaging/rdkafka.js:111]: Thanks!

File: package-lock.json
Comment (kirrg001) [reply package-lock.json:107]: Looks great now :) Thanks

File: packages/collector/test/tracing/messaging/node-rdkafka/producer.js
Comment (basti1302) [inline packages/collector/test/tracing/messaging/node-rdkafka/producer.js:49]: Should we also have a minimal test case where objectMode is false?

Of course, there would be no trace correlation but we can at least verify that we do not break anything when the option is set to false. The expectations for that test case would be that and exit and an entry are produced, but they would not be part of the same trace (different trace IDs).

File: package.json
Comment (basti1302) [reply package.json:175]: TBH, I'm 50/50 on including the kafka-avro tests. We (well, Willian) have invested time to build it. kafka-avro also was the request that lead to the instrumentation of node-rdkafka. OTOH, a "significant slowdown" (as Willian put) it plus the unfixable audit issue are clear downsides. 

Either way, it is good to have tested kavka-avro with this a couple of times on CI, even if we do not merge this particular integration test into main in the end.

File: packages/core/src/tracing/instrumentation/messaging/rdkafka.js
Comment (basti1302) [reply packages/core/src/tracing/instrumentation/messaging/rdkafka.js:215]: nitpick/opinion/matter of taste: I would marginally favor an early return for the cases that Kate mentioned. 

That is, something like

```
if (!isActive) {
   return originalEmit.apply(ctx, originalArgs);
}
if (parentSpan) {
   return originalEmit.apply(ctx, originalArgs);
}
if (cls.tracingSuppressed()) {
       const headers = addTraceLevelSuppression(messageData.headers);
        messageData.headers = headers;
        return originalEmit.apply(ctx, originalArgs);
}

... rest of the instrumentation logic
```

However, really just my opinion/preference, not a blocker or anything.

File: packages/core/src/tracing/instrumentation/messaging/rdkafka.js
Comment (basti1302) [inline packages/core/src/tracing/instrumentation/messaging/rdkafka.js:216]: Hm, actually this case does not really make sense here. This is the consumer/entry span instrumentation right? How would there be a suppression flag in the cls context if we have not evaluated the headers yet? 

Normally, for entry instrumentations, the order is as follows: Check the incoming headers for trace level suppression/X_INSTANA_L=0. If that is present, set the suppression flag in CLS. Then, _after that_, directly before creating the span, check if suppression is active and create a span or not depending on the outcome.

Or am I misreading this?

File: packages/core/src/tracing/instrumentation/messaging/rdkafka.js
Comment (basti1302) [inline packages/core/src/tracing/instrumentation/messaging/rdkafka.js:268]: We should also allow for the level header not to be present at all and assume level 1 as default. This check should probably more be like `level === 0` instead of `level !== 1`.

File: packages/core/src/tracing/instrumentation/messaging/rdkafka.js
Comment (basti1302) [inline packages/core/src/tracing/instrumentation/messaging/rdkafka.js:325]: Question: Do we actually want to use `both` as default even though we know the binary headers will not be handled correctly? WDYT about defaulting to `string` directly as long as the issue in node-rdkafka is not fixed? 

OTOH, if the node-rdkafka issue gets fixed, we get working trace correlation automatically, even if the consumer on the other end only understands the binary Instana headers. Maybe `both` is actually the correct choice. 🤔

**Review by basti1302 — APPROVED** [review]
Outstanding work! I added a few minor comments and questions, but this looks great.

File: packages/collector/test/tracing/messaging/kafka-avro/producer.js
Comment (kirrg001) [reply packages/collector/test/tracing/messaging/kafka-avro/producer.js:10]: The point was not to remove something in the PR :)

How can the customer use kafka-avro when there is no general docs about headerFormat?
Refs https://github.ibm.com/instana/docs/pull/949/files

Let's discuss offline :)

File: package.json
Comment (kirrg001) [reply package.json:175]: How about we install kafka-avro manually in our test only for v14?
Similar to what we do in our [ES tests](https://github.com/instana/nodejs/blob/main/packages/collector/test/tracing/database/elasticsearch_modern/test.js#L606)

This would I guess resolve both problems.

File: packages/collector/test/tracing/messaging/node-rdkafka/test.js
Comment (kirrg001) [reply packages/collector/test/tracing/messaging/node-rdkafka/test.js]: Yeah I totally understand 

I might try to increase code coverage before we merge the PR if that is okay for you :)

File: packages/core/src/tracing/instrumentation/messaging/rdkafka.js
Comment (kirrg001) [reply packages/core/src/tracing/instrumentation/messaging/rdkafka.js:215]: I will take this over if it's okay for you @willianpc

File: package.json
Comment (kirrg001) [reply package.json:175]: I can take over this too :)

Comment (basti1302) [reply package.json:175]: > How about we install kafka-avro manually in our test only for v14?

I think this is a great idea! 👍

File: packages/collector/test/test_util/ProcessControls.js
Comment (basti1302) [inline packages/collector/test/test_util/ProcessControls.js:293]: @kirrg001 did you mean to delete line 288? Right now we would call `resolve` twice.

Comment (kirrg001) [reply packages/collector/test/test_util/ProcessControls.js:293]: Whoops 😅

File: packages/core/src/tracing/instrumentation/messaging/rdkafka.js
Comment (kirrg001) [reply packages/core/src/tracing/instrumentation/messaging/rdkafka.js]: I'll fix that cc @willianpc

File: packages/collector/test/tracing/messaging/node-rdkafka/producer.js
Comment (kirrg001) [reply packages/collector/test/tracing/messaging/node-rdkafka/producer.js:51]: Will take over too

File: packages/collector/test/tracing/messaging/kafka-avro/producer.js
Comment (willianpc) [reply packages/collector/test/tracing/messaging/kafka-avro/producer.js:10]: I will push the change from the Go side as well. The card is sitting there for a while.

File: packages/collector/test/tracing/messaging/kafka-avro/consumer.js
Comment (willianpc) [reply packages/collector/test/tracing/messaging/kafka-avro/consumer.js:31]: It makes sense. Even though I faced no issues locally or in the pipelines, but clearly this change makes sense

File: packages/collector/test/tracing/messaging/kafka-avro/test.js
Comment (willianpc) [reply packages/collector/test/tracing/messaging/kafka-avro/test.js:47]: Makes sense. the correct semver test then needs to be 10 - 14. According to this cheat sheet: https://devhints.io/semver
I also tested it in https://semver.npmjs.com/ so it does the work.

Comment (willianpc) [reply packages/collector/test/tracing/messaging/kafka-avro/test.js:47]: @kirrg001 could you please do the change to `10 - 14`?

Comment (kirrg001) [reply packages/collector/test/tracing/messaging/kafka-avro/test.js:47]: Yes 👍

File: packages/collector/test/tracing/messaging/node-rdkafka/producer.js
Comment (willianpc) [reply packages/collector/test/tracing/messaging/node-rdkafka/producer.js:51]: good catch. I missed that one

File: packages/core/src/tracing/instrumentation/messaging/rdkafka.js
Comment (willianpc) [reply packages/core/src/tracing/instrumentation/messaging/rdkafka.js]: Oh no, you are right. I must have copied/pasted this from somewhere else and missed it.
Thanks for pointing it out 🥇

File: packages/core/src/tracing/instrumentation/messaging/rdkafka.js
Comment (willianpc) [reply packages/core/src/tracing/instrumentation/messaging/rdkafka.js]: Good point!

File: packages/core/src/tracing/instrumentation/messaging/rdkafka.js
Comment (kirrg001) [reply packages/core/src/tracing/instrumentation/messaging/rdkafka.js]: I'll take it over 👍

File: packages/core/src/tracing/instrumentation/messaging/rdkafka.js
Comment (willianpc) [reply packages/core/src/tracing/instrumentation/messaging/rdkafka.js:328]: Trying to think from a customer's perspective, I think keeping "both" sounds correct.
As you said, if node-rdkafka is fixed, the default option of "both" becomes true again.
And we have the node-rdkafka current limitation already documented in the docs.

My 2 cents: keep it as is. 
If you folks decide to change it, we just need some sort of reminder to change it back when node-rdkafka is updated with the fix. And even then, there may be the need of a version validation to switch the default behavior.

File: packages/collector/test/tracing/messaging/kafka-avro/consumer.js
Comment (kirrg001) [reply packages/collector/test/tracing/messaging/kafka-avro/consumer.js:31]: Will change

File: packages/core/src/tracing/instrumentation/messaging/rdkafka.js
Comment (basti1302) [reply packages/core/src/tracing/instrumentation/messaging/rdkafka.js:328]: Yes, that makes sense. Thanks for pointing this out.

Comment (kirrg001) [reply packages/core/src/tracing/instrumentation/messaging/rdkafka.js:312]: Summary: 

- currently is "binary"
- we want to switch to "both" for rdkafka because customer does not need to set "headerFormat" explicitly
- "headerFormat" is not even documented yet
- using "both" does not hurt, because we just send two versions of the header format

**Review by kirrg001 — APPROVED** [review]
Done 🦾

**Pull request merged**
